File: Laba1-Python/Task6.py
import Extensions

def Run():
    print("Завдання 6. Написати програму яка обраховує атаки шахматних фігур.")

    chess1_pos = 0, 0
    chess2_pos = 0, 0
    chess3_pos = 0, 0

    chess1_type = "Ферзь"
    chess2_type = "Король"
    chess3_type = "Ферзь"

    chess1_color = "Білий"
    chess2_color = "Чорний"
    chess3_color = "Чорний"

    chess1_pos = InputChessPosition(chess1_type, chess1_color)
    chess2_pos = InputChessPosition(chess2_type, chess2_color)
    chess3_pos = InputChessPosition(chess3_type, chess3_color)

    CalculateChessAttacks(chess1_pos, chess2_pos, chess3_pos,  # chess 1 attack
                          chess1_type, chess2_type, chess3_type,
                          chess1_color, chess2_color, chess3_color)
    CalculateChessAttacks(chess2_pos, chess1_pos, chess3_pos,  # chess 2 attack
                          chess2_type, chess1_type, chess3_type,
                          chess2_color, chess1_color, chess3_color)
    CalculateChessAttacks(chess3_pos, chess2_pos, chess1_pos,  # chess 3 attack
                          chess3_type, chess2_type, chess1_type,
                          chess3_color, chess2_color, chess1_color)


def IsQueenAttackingChess(queen_pos, attacked_chess_pos):
    return queen_pos[0] == attacked_chess_pos[0] or queen_pos[1] == attacked_chess_pos[1] or \
           abs(queen_pos[0] - attacked_chess_pos[0]) == abs(queen_pos[1] - attacked_chess_pos[1])

def IsKingAttackingChess(king_pos, attacked_chess_pos):
    delta_x = abs(king_pos[0] - attacked_chess_pos[0])
    delta_y = abs(king_pos[1] - attacked_chess_pos[1])
    return delta_x <= 1 and delta_y <= 1


def CalculateChessAttacks(attacking_chess_pos, chess2_pos, chess3_pos,
                          attacking_chess_type, chess2_type, chess3_type,
                          attacking_chess_color, chess2_color, chess3_color):

    if attacking_chess_type == "Ферзь":
        if attacking_chess_color != chess2_color and IsQueenAttackingChess(attacking_chess_pos, chess2_pos):
            print(f"{attacking_chess_color} ферзь атакує {chess2_color} {chess2_type}")

        if attacking_chess_color != chess3_color and IsQueenAttackingChess(attacking_chess_pos, chess3_pos):
            print(f"{attacking_chess_color} ферзь атакує {chess3_color} {chess3_type}")

    elif attacking_chess_type == "Король":
        if attacking_chess_color != chess2_color and IsKingAttackingChess(attacking_chess_pos, chess2_pos):
            print(f"{attacking_chess_color} король атакує {chess2_color} {chess2_type}")

        if attacking_chess_color != chess3_color and IsKingAttackingChess(attacking_chess_pos, chess3_pos):
            print(f"{attacking_chess_color} король атакує {chess3_color} {chess3_type}")

def InputChessPosition(type, color):
    print(f"Введіть позицію {color} {type}")

    x = Extensions.SafeIntInput("\tX")
    y = Extensions.SafeIntInput("\tY")

    return x, y

File: Laba1/Task6.cs
using Extensions;
using System.Drawing;

namespace PZPK_Labas.Laba1
{
    public static class Task6
    {
        public static void Run()
        {
            Console.WriteLine("Завдання 6. Написати програму яка обраховує атаки шахматних фігур.");

            Point chess1Pos;
            Point chess2Pos;
            Point chess3Pos;

            string chess1Type = "Ферзь";
            string chess2Type = "Король";
            string chess3Type = "Ферзь";

            string chess1Color = "Білий";
            string chess2Color = "Чорний";
            string chess3Color = "Чорний";

            chess1Pos = InputChessPosition(chess1Type, chess1Color);
            chess2Pos = InputChessPosition(chess2Type, chess2Color);
            chess3Pos = InputChessPosition(chess3Type, chess3Color);

            CalculateChessAttacks(chess1Pos, chess2Pos, chess3Pos, // chess 1 attack
                                    chess1Type, chess2Type, chess3Type,
                                    chess1Color, chess2Color, chess3Color);
            CalculateChessAttacks(chess2Pos, chess1Pos, chess3Pos, // chess 2 attack
                                    chess2Type, chess1Type, chess3Type,
                                    chess2Color, chess1Color, chess3Color);
            CalculateChessAttacks(chess3Pos, chess2Pos, chess1Pos, // chess 3 attack
                                    chess3Type, chess2Type, chess1Type,
                                    chess3Color, chess2Color, chess1Color);

            Console.ReadKey();
            Console.WriteLine();
        }

        private static Point InputChessPosition(string type, string color)
        {
            Console.WriteLine($"Введіть позицію {color} {type}");

            int x;
            int y;

            ConsoleInput.SafeIntInput("\tX", out x);
            ConsoleInput.SafeIntInput("\tY", out y);

            return new Point(x, y);
        }

        private static void CalculateChessAttacks(Point attackingChessPos, Point chess2Pos, Point chess3Pos,
                                                    string attackingChessType, string chess2Type, string chess3Type,
                                                    string attachingChessColor, string chess2Color, string chess3Color)
        {
            switch (attackingChessType)
            {
                case "Ферзь":
                    if (attachingChessColor != chess2Color & IsQueenAttackingChess(attackingChessPos, chess2Pos))
                        Console.WriteLine($"{attachingChessColor} ферзь атакує {chess2Color} {chess2Type}");

                    if (attachingChessColor != chess3Color & IsQueenAttackingChess(attackingChessPos, chess3Pos))
                        Console.WriteLine($"{attachingChessColor} ферзь атакує {chess3Color} {chess3Type}");
                    break;
                case "Король":
                    if (attachingChessColor != chess2Color & IsKingAttackingChess(attackingChessPos, chess2Pos))
                        Console.WriteLine($"{attachingChessColor} король атакує {chess2Color} {chess2Type}");

                    if (attachingChessColor != chess3Color & IsKingAttackingChess(attackingChessPos, chess3Pos))
                        Console.WriteLine($"{attachingChessColor} король атакує {chess3Color} {chess3Type}");
                    break;
            }
        }

        private static bool IsQueenAttackingChess(Point queenPos, Point attackedChessPos)
        {
            return queenPos.X == attackedChessPos.X || queenPos.Y == attackedChessPos.Y ||
                    Math.Abs(queenPos.X - attackedChessPos.X) == Math.Abs(queenPos.Y - attackedChessPos.Y);
        }

        private static bool IsKingAttackingChess(Point kingPos, Point attackedChessPos)
        {
            int deltaX = Math.Abs(kingPos.X - attackedChessPos.X);
            int deltaY = Math.Abs(kingPos.Y - attackedChessPos.Y);

            return deltaX <= 1 & deltaY <= 1;
        }
    }
}